README.md
# NAME

DBIx::Class::Visualizer - Visualize a DBIx::Class schema

<div>
    <p>
    <img src="https://img.shields.io/badge/perl-5.10.1+-blue.svg" alt="Requires Perl 5.10.1+" />
    <a href="https://travis-ci.org/Csson/p5-DBIx-Class-Visualizer"><img src="https://api.travis-ci.org/Csson/p5-DBIx-Class-Visualizer.svg?branch=master" alt="Travis status" /></a>
    <a href="http://cpants.cpanauthors.org/release/CSSON/DBIx-Class-Visualizer-0.0200"><img src="http://badgedepot.code301.com/badge/kwalitee/CSSON/DBIx-Class-Visualizer/0.0200" alt="Distribution kwalitee" /></a>
    <a href="http://matrix.cpantesters.org/?dist=DBIx-Class-Visualizer%200.0200"><img src="http://badgedepot.code301.com/badge/cpantesters/DBIx-Class-Visualizer/0.0200" alt="CPAN Testers result" /></a>
    <img src="https://img.shields.io/badge/coverage-89.2%-orange.svg" alt="coverage 89.2%" />
    </p>
</div>

# VERSION

Version 0.0200, released 2016-09-19.

# SYNOPSIS

    use DBIx::Class::Visualizer;
    use The::DBIxClassSchema;

    my $schema = The::DBIxClassSchema->connect;
    my $svg = DBIx::Class::Visualizer->new(schema => $schema)->svg;

# DESCRIPTION

DBIx::Class::Visualizer is a [GraphViz2](https://metacpan.org/pod/GraphViz2) renderer for [DBIx::Class](https://metacpan.org/pod/DBIx::Class) schemas. It is designed to be used as a backend to web applications that
can display the rendered graph in a more user friendly way. See [Mojolicious::Plugin::DbicSchemaViewer](https://metacpan.org/pod/Mojolicious::Plugin::DbicSchemaViewer).

# STATUS

Backwards compatability between even minor releases is currently not a goal. That said, the public interface is small and most
breaking changes are likely to be in ["transformed\_svg"](#transformed_svg).

# ATTRIBUTES

## schema

Required. An instance of a [DBIx::Class::Schema](https://metacpan.org/pod/DBIx::Class::Schema) class.

## logger\_conf

Optional array reference. [GraphViz2](https://metacpan.org/pod/GraphViz2) uses [Log::Handler](https://metacpan.org/pod/Log::Handler), so this distribution does that too. By default it noisily prints to screen. Not used if `logger` is set.

## logger

Optional. An instance of [Log::Handler](https://metacpan.org/pod/Log::Handler).

## wanted\_result\_source\_names

Optional. An array reference consisting of result source names (without the .\*::Result:: prefix) you wish to include in the output. This can
be useful to focus on a small part of large schemas.

If it is not set all result sources will be rendered (minus ["skip\_result\_source\_names"](#skip_result_source_names)).

## skip\_result\_source\_names

Optional. An array reference consisting of result source names (without the .\*::Result:: prefix) you wish to not include in the output.

## degrees\_of\_separation

Optional. A non-negative integer that is used together with ["wanted\_result\_source\_names"](#wanted_result_source_names). In addition to the wanted result sources, this attribute
defines how many relationship steps should be followed to other result sources that also should be included in the output.

Default is `1`.

## only\_keys

Boolean, defaults to `0`. If true, only primary and foreign key columns will be rendered.

## graphviz\_conf

Optional hashref. This hashref is passed to the [GraphViz2](https://metacpan.org/pod/GraphViz2) constructor. The output from ["transformed\_svg"](#transformed_svg) is adapted to the default settings, so
using these two together might cause a less usable svg document.

Won't be used if you pass `graph` to the constructor.

## graph

Optional. A [GraphViz2](https://metacpan.org/pod/GraphViz2) object. Pass this if you need to use an already constructed graph.

After ["new"](#new) has run it can be useful if you, for example, wishes to see the arguments to the dot renderer:

    my $visualizer = DBIx::Class::Visualizer->new(schema => $schema);
    my $svg = $visualizer->svg;

    my $dotfile = $visualizer->graph->dot_input;

# METHODS

## new

The constructor.

## svg

Takes no arguments, and returns the graph as an svg string.

## run

A shortcut for ["run" in GraphViz2](https://metacpan.org/pod/GraphViz2#run):

    DBIx::Class::Visualizer->new(schema => $schema)->run(output_file => 'myschema.png', format => 'png');

## transformed\_svg

Takes no arguments. Returns an svg string that is more useful than that from ["svg"](#svg) Using this method requires [Mojolicious](https://metacpan.org/pod/Mojolicious).

This method improves the svg generated by [graphviz](http://graphviz.org/) in several ways:

- All layout attributes (eg. `fill`, `stroke`, `font-family`) are removed so that styling can be done using css.
- There are occasional minor gaps between the various elements in edges, these are removed (or at least reduced).
- This distribution adds some padding between texts and borders to avoid overlapping. These are removed so that no unnecessary elements remain.
- All edges, nodes and column name elements get relevant values for their id attributes.
- Several `data-` attributes are added to edges, nodes and column attributes containing a lot of information about the schema.

As an example, this is a column element as rendered by `graphviz` (whitespace added for readability:

    <text text-anchor="start"
           x="700.391"
           y="-17.9"
           font-family="Helvetica,sans-Serif"
           font-weight="bold"
           font-size="10.00"
           fill="#222222">a_column_id</text>

After passing through `transformed_svg` the same column looks like this:

    <text id="column-TableName-a_column_id"
          class="column-name"
          y="-17.9"
          x="700.391"
          text-anchor="start"
          data-is-primary="1"
          data-column-name="a_column_id"
          data-column-info="{
            "name": "a_column_id",
            "data_type": "integer",
            "is_primary_key": 1
            "is_auto_increment": 1,
            "is_nullable": 0,
            "is_foreign_key": 0,
            "is_numeric": 1,
            "extra": {},
            "relations":[
                {
                    "origin_table": "TableName",
                    "origin_column": "a_column_id",
                    "destination_table": "AnotherTableName"
                    "destination_column": "a_column_id",
                    "relation_type": "has_many",
                    "cascade_delete": 1,
                },
                ...
            ],
          }">a_column_id</text>

The `data-column-info` attribute is a json object that is directly usable by something like jQuery:

    # has_many
    $('#column-TableName-a_column_id').data('column-info').relations[0].relation_type;

# SEE ALSO

- [Mojolicious::Plugin::DbicSchemaViewer](https://metacpan.org/pod/Mojolicious::Plugin::DbicSchemaViewer) - A [Mojolicious](https://metacpan.org/pod/Mojolicious) plugin that uses this class
- [GraphViz2::DBI](https://metacpan.org/pod/GraphViz2::DBI) - Visualizes a schema by inspecting the database.

# SOURCE

[https://github.com/Csson/p5-DBIx-Class-Visualizer](https://github.com/Csson/p5-DBIx-Class-Visualizer)

# HOMEPAGE

[https://metacpan.org/release/DBIx-Class-Visualizer](https://metacpan.org/release/DBIx-Class-Visualizer)

# AUTHOR

Erik Carlsson <info@code301.com>

# COPYRIGHT AND LICENSE

This software is copyright (c) 2016 by Erik Carlsson.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Revision history for DBIx-Class-Visualizer

{{$NEXT}}
 [API Changes]

 [Bug Fixes]

 [Documentation]

 [Enhancements]

 [New Features]

 [Other Changes]

 [Requirements]

 [Stats & Dependency Changes]

 [Testing]

0.0200 2016-09-19T23:33:53Z UTC
 [Bug Fixes]
 - Remove use of /r in a regexp.

 [Documentation]
 - Removed outdated example, but Mojolicious::Plugin::DbicSchemaViewer has
   a demo.

 [New Features]
 - Can transform the rendered svg document so that it is much more useful
   in a web application.

 [Requirements]
 - Mojolicious is required to make use of the improved svg transformation
   (but not otherwise).

 [Stats & Dependency Changes]
 - 13 files changed, 1154 insertions(+), 271 deletions(-)
 - (run req) + JSON::MaybeXS 1.003003
 - (run req) + Log::Handler 0.84
 - (run req) + PerlX::Maybe 1.001
 - (run req) + ReadonlyX 1.03
 - (run req) + Syntax::Keyword::Gather 1.002000
 - (run req) + Types::Standard 1.000005
 - (run req) ~ GraphViz2 2.38 --> 2.42
 - (run rec) + Mojolicious 6.0
 - (test req) + DBIx::Class 0.08260

0.0100 2016-09-04T20:13:20Z UTC
 - Initial release

 [Stats & Dependency Changes]
 - (run req) + DateTime::Tiny 1.06
 - (run req) + GraphViz2 2.38
 - (run req) + List::Util 1.45
 - (run req) + Moo 2.00
 - (run req) + perl 5.010001
 - (test req) + ExtUtils::MakeMaker (any)
 - (test req) + File::Spec (any)
 - (test req) + IO::Handle (any)
 - (test req) + IPC::Open3 (any)
 - (test req) + Test::More 0.96
 - (test rec) + CPAN::Meta 2.120900
 - (conf req) + ExtUtils::MakeMaker (any)
 - (dev req) + Badge::Depot 0.0103
 - (dev req) + Badge::Depot::Plugin::Coverage (any)
 - (dev req) + Badge::Depot::Plugin::Cpantesters (any)
 - (dev req) + Badge::Depot::Plugin::Kwalitee (any)
 - (dev req) + Badge::Depot::Plugin::Perl (any)
 - (dev req) + Badge::Depot::Plugin::Travis (any)
 - (dev req) + Dist::Zilla 5.000
 - (dev req) + Dist::Zilla::Plugin::Authority 1.009
 - (dev req) + Dist::Zilla::Plugin::BumpVersionAfterRelease::Transitional
   0.007
 - (dev req) + Dist::Zilla::Plugin::ChangeStats::Dependencies::Git 0.0100
 - (dev req) + Dist::Zilla::Plugin::CheckChangesHasContent 0.008
 - (dev req) + Dist::Zilla::Plugin::Clean 0.07
 - (dev req) + Dist::Zilla::Plugin::ConfirmRelease (any)
 - (dev req) + Dist::Zilla::Plugin::CopyFilesFromBuild 0.151680
 - (dev req) + Dist::Zilla::Plugin::DistIller::MetaGeneratedBy (any)
 - (dev req) + Dist::Zilla::Plugin::ExecDir (any)
 - (dev req) + Dist::Zilla::Plugin::Git 2.036
 - (dev req) + Dist::Zilla::Plugin::Git::Check (any)
 - (dev req) + Dist::Zilla::Plugin::Git::CheckFor::CorrectBranch 0.013
 - (dev req) + Dist::Zilla::Plugin::Git::Commit (any)
 - (dev req) + Dist::Zilla::Plugin::Git::Contributors 0.017
 - (dev req) + Dist::Zilla::Plugin::Git::GatherDir (any)
 - (dev req) + Dist::Zilla::Plugin::Git::Push (any)
 - (dev req) + Dist::Zilla::Plugin::Git::Tag (any)
 - (dev req) + Dist::Zilla::Plugin::GithubMeta 0.54
 - (dev req) + Dist::Zilla::Plugin::InstallRelease 0.008
 - (dev req) + Dist::Zilla::Plugin::License (any)
 - (dev req) + Dist::Zilla::Plugin::MakeMaker (any)
 - (dev req) + Dist::Zilla::Plugin::Manifest (any)
 - (dev req) + Dist::Zilla::Plugin::ManifestSkip (any)
 - (dev req) + Dist::Zilla::Plugin::MetaConfig (any)
 - (dev req) + Dist::Zilla::Plugin::MetaJSON (any)
 - (dev req) + Dist::Zilla::Plugin::MetaNoIndex (any)
 - (dev req) + Dist::Zilla::Plugin::MetaProvides::Package 2.003001
 - (dev req) + Dist::Zilla::Plugin::MetaYAML (any)
 - (dev req) + Dist::Zilla::Plugin::NextRelease::Grouped 0.0200
 - (dev req) + Dist::Zilla::Plugin::PodSyntaxTests (any)
 - (dev req) + Dist::Zilla::Plugin::PodWeaver 4.006
 - (dev req) + Dist::Zilla::Plugin::Prereqs (any)
 - (dev req) + Dist::Zilla::Plugin::PromptIfStale (any)
 - (dev req) + Dist::Zilla::Plugin::Readme (any)
 - (dev req) + Dist::Zilla::Plugin::ReadmeAnyFromPod 0.150250
 - (dev req) + Dist::Zilla::Plugin::RewriteVersion::Transitional 0.007
 - (dev req) + Dist::Zilla::Plugin::RunExtraTests 0.028
 - (dev req) + Dist::Zilla::Plugin::ShareDir (any)
 - (dev req) + Dist::Zilla::Plugin::Test::Compile 2.054
 - (dev req) + Dist::Zilla::Plugin::Test::EOF 0.0501
 - (dev req) + Dist::Zilla::Plugin::Test::EOL 0.18
 - (dev req) + Dist::Zilla::Plugin::Test::Kwalitee::Extra v0.2.1
 - (dev req) + Dist::Zilla::Plugin::Test::NoTabs 0.15
 - (dev req) + Dist::Zilla::Plugin::Test::ReportPrereqs 0.021
 - (dev req) + Dist::Zilla::Plugin::Test::Version 1.05
 - (dev req) + Dist::Zilla::Plugin::TestRelease (any)
 - (dev req) + Dist::Zilla::Plugin::TravisYML 1.15
 - (dev req) + Dist::Zilla::Plugin::UploadToCPAN (any)
 - (dev req) + Pod::Elemental::Transformer::List (any)
 - (dev req) + Pod::Weaver::Plugin::SingleEncoding (any)
 - (dev req) + Pod::Weaver::Plugin::Transformer (any)
 - (dev req) + Pod::Weaver::PluginBundle::CorePrep (any)
 - (dev req) + Pod::Weaver::Section::Authors (any)
 - (dev req) + Pod::Weaver::Section::Badges 0.0401
 - (dev req) + Pod::Weaver::Section::Collect (any)
 - (dev req) + Pod::Weaver::Section::GenerateSection 1.01
 - (dev req) + Pod::Weaver::Section::Generic (any)
 - (dev req) + Pod::Weaver::Section::Leftovers (any)
 - (dev req) + Pod::Weaver::Section::Legal (any)
 - (dev req) + Pod::Weaver::Section::Name (any)
 - (dev req) + Pod::Weaver::Section::Region (any)
 - (dev req) + Pod::Weaver::Section::Version (any)
 - (dev req) + Test::EOF (any)
 - (dev req) + Test::EOL (any)
 - (dev req) + Test::More 0.88
 - (dev req) + Test::NoTabs 1.4
 - (dev req) + Test::Pod 1.41
 - (dev req) + Test::Version 1
 - (dev req) + Test::Warnings 0.026
 - (dev sug) + Dist::Iller 0.1408
 - (dev sug) + Dist::Iller::Config::Author::CSSON 0.0312


lib/DBIx/Class/Visualizer/Column.pm
use 5.10.1;
use strict;
use warnings;

package DBIx::Class::Visualizer::Column;

# ABSTRACT: Handle column information
# AUTHORITY
our $VERSION = '0.0201';

use Moo;
use DBIx::Class::Visualizer::Relation;
use Types::Standard qw/ArrayRef InstanceOf/;
use PerlX::Maybe qw/provided/;

has name => (
    is => 'ro',
    required => 1,
);
has data_type => (
    is => 'ro',
    required => 1,
);
has relations => (
    is => 'ro',
    isa => ArrayRef[InstanceOf['DBIx::Class::Visualizer::Relation']],
    lazy => 1,
    default => sub { [] },
);

has is_primary_key => (
    is => 'ro',
    default => 0,
);
has is_foreign_key => (
    is => 'ro',
    default => 0,
);
has is_nullable => (
    is => 'ro',
    default => 0,
);

has accessor => (
    is => 'ro',
    predicate => 1,
);
has size => (
    is => 'ro',
    predicate => 1,
);
has is_auto_increment => (
    is => 'ro',
    predicate => 1,
);
has is_numeric => (
    is => 'ro',
    predicate => 1,
);
has default_value => (
    is => 'ro',
    predicate => 1,
);
has sequence => (
    is => 'ro',
    predicate => 1,
);
has retrieve_on_insert => (
    is => 'ro',
    predicate => 1,
);
has unsigned => (
    is => 'ro',
    predicate => 1,
);
has extra => (
    is => 'ro',
    predicate => 1,
);

around BUILDARGS => sub {
    my $orig = shift;
    my $class = shift;
    my %args = @_;

    if(exists $args{'extra'}{'unsigned'}) {
        $args{'unsigned'} = delete $args{'extra'}{'unsigned'};
    }
    # Dereference default values like \'now()'
    if(exists $args{'default_value'}) {
        if(ref $args{'default_value'} eq 'SCALAR') {
            $args{'default_value'} = ${ $args{'default_value'} };
        }
        # ..and put quotes around string default values
        else {
            $args{'default_value'} = qq{'$args{'default_value'}'};
        }
    }

    # remove things like _inflate_info
    map { delete $args{ $_ } } grep { /^_/ } keys %args;

    $class->$orig(%args);
};

sub column_name_label_tag {
    my $self = shift;

    my $column_name_tag = $self->name;
    $column_name_tag = $self->{'is_primary_key'} ? "<b>$column_name_tag</b>" : $column_name_tag;
    $column_name_tag = $self->{'is_foreign_key'} ? "<u>$column_name_tag</u>" : $column_name_tag;
    return $column_name_tag;
}

sub TO_JSON {
    my $self = shift;

    return +{
        name => $self->name,
        data_type => $self->data_type,
        relations => [map { $_->TO_JSON } @{ $self->relations }],
        is_primary_key => $self->is_primary_key,
        is_foreign_key => $self->is_foreign_key,
        is_nullable => $self->is_nullable,
        provided $self->has_accessor,           accessor => $self->accessor,
        provided $self->has_size,               size => $self->size,
        provided $self->has_is_auto_increment,  is_auto_increment => $self->is_auto_increment,
        provided $self->has_is_numeric,         is_numeric => $self->is_numeric,
        provided $self->has_default_value,      default_value => $self->default_value,
        provided $self->has_sequence,           sequence => $self->sequence,
        provided $self->has_retrieve_on_insert, retrieve_on_insert => $self->retrieve_on_insert,
        provided $self->has_unsigned,           unsigned => $self->unsigned,
        provided $self->has_extra,              extra => $self->extra,
    };
}

1;


lib/DBIx/Class/Visualizer/Relation.pm
use 5.10.1;
use strict;
use warnings;

package DBIx::Class::Visualizer::Relation;

# ABSTRACT: Handle relation information
# AUTHORITY
our $VERSION = '0.0201';

use Moo;
use ReadonlyX;
use Types::Standard qw/Str Bool Enum/;
use PerlX::Maybe qw/provided/;

Readonly::Scalar my $HAS_MANY => 'has_many';
Readonly::Scalar my $BELONGS_TO => 'belongs_to';
Readonly::Scalar my $HAS_ONE => 'has_one';
Readonly::Scalar my $MIGHT_HAVE => 'might_have';

has added_to_graph => (
    is => 'rw',
    isa => Bool,
    default => 0,
);

has origin_table => (
    is => 'ro',
    isa => Str,
    required => 1,
);
has origin_column => (
    is => 'ro',
    isa => Str,
    required => 1,
);
has destination_table => (
    is => 'ro',
    isa => Str,
    required => 1,
);
has destination_column => (
    is => 'ro',
    isa => Str,
    required => 1,
);
has cascade_delete => (
    is => 'ro',
    isa => Bool,
    predicate => 1,
);
has cascade_update => (
    is => 'ro',
    isa => Bool,
    predicate => 1,
);
has relation_type => (
    is => 'ro',
    isa => Enum[qw/has_many has_one belongs_to might_have/],
    required => 1,
);

around BUILDARGS => sub {
    my $orig = shift;
    my $class = shift;
    my %args = @_;

    my $relation = delete $args{'relation'};
    my $attr = $relation->{'attrs'};

    ($args{'destination_table'} = $relation->{'source'}) =~ s{^.*?::Result::}{};
    ($args{'origin_column'} = (values %{ $relation->{'cond'} })[0]) =~ s{^self\.}{};
    ($args{'destination_column'} = (keys %{ $relation->{'cond'} })[0]) =~ s{^foreign\.}{};

    for my $cascade (qw/cascade_delete cascade_update/) {
        $args{ $cascade } = $attr->{ $cascade } if exists $attr->{ $cascade };
    }

    # do not reorder
    $args{'relation_type'} = $attr->{'accessor'} eq 'multi' ? $HAS_MANY
                           : $attr->{'is_depends_on'}       ? $BELONGS_TO
                           : exists $attr->{'join_type'}    ? $MIGHT_HAVE
                           :                                  $HAS_ONE
                           ;

    $class->$orig(%args);
};

sub is_belongs_to { shift->relation_type eq $BELONGS_TO }

sub arrow_type {
    my $self = shift;

    return join '', $self->relation_type eq $HAS_MANY   ? qw/crow none odot/
                  : $self->relation_type eq $BELONGS_TO ? qw/none tee/
                  : $self->relation_type eq $MIGHT_HAVE ? qw/none tee none odot/
                  : $self->relation_type eq $HAS_ONE    ? qw/vee/
                  :                                       qw/dot dot dot/
                  ;
}

sub TO_JSON {
    my $self = shift;

    return +{
            origin_table => $self->origin_table,
            origin_column => $self->origin_column,
            destination_table => $self->destination_table,
            destination_column => $self->destination_column,
            relation_type => $self->relation_type,
        provided $self->has_cascade_delete,
            cascade_delete => $self->cascade_delete,
        provided $self->has_cascade_update,
            cascade_update => $self->cascade_update,
    };
}

1;


lib/DBIx/Class/Visualizer/ResultHandler.pm
use 5.10.1;
use strict;
use warnings;

package DBIx::Class::Visualizer::ResultHandler;

# ABSTRACT: Handle result sources and related information
# AUTHORITY
our $VERSION = '0.0201';

use Moo;
use Types::Standard qw/Bool ArrayRef HashRef InstanceOf Int/;
use PerlX::Maybe;
use List::Util qw/any/;
use Syntax::Keyword::Gather;
use DBIx::Class::Visualizer::Relation;
use DBIx::Class::Visualizer::Column;

has name => (
    is => 'ro',
    required => 1,
);
has rs => (
    is => 'ro',
    required => 1,
);
has show => (
    is => 'rw',
    isa => Bool,
    default => 1,
);
has wanted => (
    is => 'ro',
    isa => Bool,
    default => 0,
);
has skip => (
    is => 'ro',
    isa => Bool,
    default => 0,
);
has columns => (
    is => 'ro',
    isa => ArrayRef[InstanceOf['DBIx::Class::Visualizer::Column']],
    lazy => 1,
    builder => 1,
);
has only_keys => (
    is => 'ro',
    isa => Bool,
    default => 0,
);
has relations => (
    is => 'ro',
    isa => ArrayRef[InstanceOf['DBIx::Class::Visualizer::Relation']],
    lazy => 1,
    builder => 1,
);
has _degree_of_separation => (
    is => 'rw',
    isa => Int,
    init_arg => 'degree_of_separation',
    predicate => 1,
);
sub degree_of_separation {
    my $self = shift;
    my $value = shift;
    return 0 if !defined $value && !$self->_has_degree_of_separation;
    return $self->_degree_of_separation if $self->_has_degree_of_separation && !defined $value;
    return $self->_degree_of_separation if $self->_has_degree_of_separation && $value > $self->_degree_of_separation;
    $self->_degree_of_separation($value);
}

sub _build_columns {
    my $self = shift;

    my @primary_columns = $self->rs->primary_columns;

    return [
        gather {
            COLUMN:
            for my $column_name ($self->rs->columns) {

                my $column_info = $self->rs->column_info($column_name);
                my $is_primary = (any { $column_name eq $_ } @primary_columns) ? 1 : 0;
                my $is_foreign = $column_info->{'is_foreign_key'} ? 1 : 0;

                next COLUMN if !$is_primary && !$is_foreign && $self->only_keys;

                take(DBIx::Class::Visualizer::Column->new(
                    name => $column_name,
                    %{ $column_info },
                    is_primary_key => $is_primary,
                    relations => [$self->get_relations($column_name)],
                ));
             }
        }
    ];
}

sub _build_relations {
    my $self = shift;

    my @relationship_names = $self->rs->relationships;

    return [
        gather {
            RELATION:
            for my $relation_name (@relationship_names) {
                my $relation = $self->rs->relationship_info($relation_name);
                next RELATION if ref $relation->{'cond'} ne 'HASH' || scalar keys %{ $relation->{'cond'} } != 1;

                take(DBIx::Class::Visualizer::Relation->new(origin_table => $self->name, relation => $relation));
            }
        }
    ];
}

sub get_relations {
    my $self = shift;
    my $column_name = shift;

    return grep { $_->origin_column eq $column_name } @{ $self->relations };
}

# find the relation from $origin_column in this result_source that points to $destination_table.$destination_column
sub get_relation_between {
    my $self = shift;
    my $origin_column = shift;
    my $destination_table = shift;
    my $destination_column = shift;

    return (grep { $_->destination_table eq $destination_table && $_->destination_column eq $destination_column } $self->get_relations($origin_column))[0];
}
sub get_column {
    my $self = shift;
    my $column_name = shift;
    return (grep { $column_name eq $_->name } @{ $self->columns })[0];
}

# Node names can't have colons in them.
sub node_name {
    my $node_name = shift->name;
    $node_name =~ s{::}{__}g;
    return $node_name;
}

1;


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.007.
use strict;
use warnings;

use 5.010001;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Visualize a DBIx::Class schema",
  "AUTHOR" => "Erik Carlsson <info\@code301.com>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "DBIx-Class-Visualizer",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.010001",
  "NAME" => "DBIx::Class::Visualizer",
  "PREREQ_PM" => {
    "DateTime::Tiny" => "1.06",
    "GraphViz2" => "2.42",
    "JSON::MaybeXS" => "1.003003",
    "List::Util" => "1.45",
    "Log::Handler" => "0.84",
    "Moo" => "2.00",
    "PerlX::Maybe" => "1.001",
    "ReadonlyX" => "1.03",
    "Syntax::Keyword::Gather" => "1.002000",
    "Types::Standard" => "1.000005"
  },
  "TEST_REQUIRES" => {
    "DBIx::Class" => "0.08260",
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::More" => "0.96"
  },
  "VERSION" => "0.0201",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "DBIx::Class" => "0.08260",
  "DateTime::Tiny" => "1.06",
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "GraphViz2" => "2.42",
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "JSON::MaybeXS" => "1.003003",
  "List::Util" => "1.45",
  "Log::Handler" => "0.84",
  "Moo" => "2.00",
  "PerlX::Maybe" => "1.001",
  "ReadonlyX" => "1.03",
  "Syntax::Keyword::Gather" => "1.002000",
  "Test::More" => "0.96",
  "Types::Standard" => "1.000005"
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
